common/range-map/src/range_map.rs

```diff
@@ -0,0 +1,151 @@
+    // Wrap ranges so that they are `Ord`.
+    // See `range_wrapper.rs` for explanation.
+    pub(crate) btm: BTreeMap<RangeWrapper<K>, V>,
+}
+
+impl<K, V> Default for RangeMap<K, V>
+where
+    K: Ord + Clone + std::fmt::Debug + Copy,
+    V: Eq + Clone + std::fmt::Debug,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct Iter<'a, K, V> {
+    inner: std::collections::btree_map::Iter<'a, RangeWrapper<K>, V>,
+}
+
+impl<'a, K, V> Iterator for Iter<'a, K, V>
+where
+    K: 'a,
+    V: 'a,
+{
+    type Item = (&'a Range<K>, &'a V);
+
+    fn next(&mut self) -> Option<(&'a Range<K>, &'a V)> {
+        self.inner.next().map(|(by_start, v)| (&by_start.range, v))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+impl<K: Debug, V: Debug> Debug for RangeMap<K, V>
+where
+    K: Ord + Clone + std::fmt::Debug + Copy,
+    V: Eq + Clone + std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
+    }
+}
+
+impl<K, V> RangeMap<K, V>
+where
+    K: Ord + Clone + std::fmt::Debug + Copy,
+    V: Eq + Clone + std::fmt::Debug,
+{
+    /// Makes a new empty `RangeMap`.
+    pub fn new() -> Self {
+        RangeMap {
+            btm: BTreeMap::new(),
+        }
+    }
+
+    pub fn contain_key(&self, key: &K) -> Vec<(&RangeWrapper<K>, &V)> {
+        let key_as_start = RangeWrapper::new(*key..*key);
+
+        let a = self
+            .btm
+            .range((Bound::Included(key_as_start.clone()), Bound::Unbounded))
+            .filter(|e| {
+                let start = e.0.range.start.cmp(&key_as_start.range.start);
+                let end = e.0.range.end.cmp(&key_as_start.range.start);
+
+                start != Ordering::Greater && end != Ordering::Less
+            })
+            .collect();
+
+        println!("k: {:?}, a: {:?}", key, a);
+        a
+    }
+
+    pub fn insert(&mut self, range: Range<K>, value: V) {
+        assert!(range.start < range.end);
+
+        let new_range_start_wrapper: RangeWrapper<K> = RangeWrapper::new(range);
+        let new_value = value;
+
+        // Insert the (possibly expanded) new range, and we're done!
+        self.btm.insert(new_range_start_wrapper, new_value);
+    }
+
+    pub fn upinsert<F>(&mut self, range: Range<K>, value: V, f: &mut F)
+    where F: for<'a> FnMut(&'a mut V, &'a V) -> V {
+        assert!(range.start < range.end);
+
+        let new_range_start_wrapper: RangeWrapper<K> = RangeWrapper::new(range);
+        let new_value = value;
+
+        let old_value = self.btm.get_mut(&new_range_start_wrapper);
+        match old_value {
+            None => {
+                self.btm.insert(new_range_start_wrapper, new_value);
+            }
+            Some(old_value) => {
+                let new_value = f(old_value, &new_value);
+                self.btm.insert(new_range_start_wrapper, new_value);
+            }
+        };
+    }
```

Comment:
To simplify the API, `F` could be a function `Fn(&Option<V>) -> V`.

This way the argument `value: V` won't be necessary. It will be the duty of `F` to produce a value if the argument is `None`.

Reply:
if pass to Fn the `Option`, then no matter or not there is existing key, the new value is decide by the function, by the design of this `upsert` function is using the fn to generate new value only when there is existing value.If there is no value exise, it will insert directly.

Reply:
Right, in case the user just wants to override the existing value, the user does not need to write the `value` twice.